### PR TITLE
Be scraping eric

### PR DESCRIPTION
**Pull request by Ethorf**
Okay so this is getting there as far as functionality! I'm going blank on getting it to return the data I need though, my Promise knowledge is not up to par quite yet, I have no problem console.log()ing the data but actually returning it from the function is eluding me. I'm guessing it's something to do with the promise resolving thing? Any and all elucidation in this area would be much appreciated!

### CODE REVIEW

File: server/utils/scraping.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		//This makes sure that the page has loaded based on one selector
+		if (domainName(url) === 'amazon') {
+			await page.waitForSelector('span#productTitle');
+
+			pageData = await page.evaluate(() => {
+				const priceBeginning = /[a-zA-Z]*/;
+				const dollarSign = /[$]/;
+				//nothing inside or outside of page.evaluate is available outside of it's closure, because it is running purely in the dom
+				let titleData = document.getElementById(`productTitle`);
+				let priceData = document.getElementById(`price_inside_buybox`);
```
Comment (surhud004) [inline server/utils/scraping.js]: If it is easier to scrape the price from `price_inside_buybox`, then no worries. Else you can try scraping from `priceblock_ourprice` too!

File: server/utils/scraping.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+				let imageData = document.getElementById(`landingImage`);
+				// let descriptionData = document.getElementById(`productDescription`);
+				// let descriptionData = [...document.querySelectorAll(`#feature-bullets ul li span`)].map(
```
Comment (surhud004) [inline server/utils/scraping.js]: One alternative to product description is - we can display `<a href="url_to_the_product" target="_blank">View Details</a>`

File: server/utils/scraping.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+					//will want to change this to just the number
+					price: priceData.innerText,
+					image: imageData.src,
+					description: descriptionData.innerText
+				};
+				return JSON.stringify(pageDataObject);
+			});
+
+			await browser.close();
+			// console.log(pageData);
```
Comment (surhud004) [inline server/utils/scraping.js]: This I found in npm puppeteer docs - So you need to return a JS object?
```
const dimensions = await page.evaluate(() => {
    return {
      width: document.documentElement.clientWidth,
      height: document.documentElement.clientHeight,
      deviceScaleFactor: window.devicePixelRatio
    };
  });
 
  console.log('Dimensions:', dimensions);
 
  await browser.close();
```

File: server/utils/scraping.js (hunk 1 above)
Comment (Ethorf) [reply server/utils/scraping.js]: yeah priceblock_ourprice was actually what I was going by first! But that ID cd changes when an item is on sale to something like "dealprice" etc. I found the buy box was more consistent when applied to the admittedly small sample size of like 10 different amazon pages.

File: server/utils/scraping.js (hunk 2 above)
Comment (Ethorf) [reply server/utils/scraping.js]: yeah I'd be down to do that!

File: server/utils/scraping.js (hunk 3 above)
Comment (Ethorf) [reply server/utils/scraping.js]: awesome I will try that!

**Comment by sina-jamshidi** [comment]
> Okay so this is getting there as far as functionality! I'm going blank on getting it to return the data I need though, my Promise knowledge is not up to par quite yet, I have no problem console.log()ing the data but actually returning it from the function is eluding me. I'm guessing it's something to do with the promise resolving thing? Any and all elucidation in this area would be much appreciated!

We can discuss this more in the standup. I wrote this [quick sandbox](https://codesandbox.io/s/tender-waterfall-59s6x?file=/src/index.js) on how Promise.Resolve() might work. Hopefully it helps. I'm not entirely sure from just looking at the code why it's not working for you

**Comment by Ethorf** [comment]
Okay so I struggled with Promises for about 3 hours today to no avail, a solution I have made right now is to have the scraping be a post route, but I'm not sure if it's unforgivably antipattern to call a route within another route, it seems like it would make more sense to have it as a function, but getting it to actually return instead of just console.log still eludes me

**Review by surhud004 — APPROVED** [review]
That looks good!!

**Pull request merged**